consensus/polybft/consensus_runtime.go

```diff
@@ -731,51 +723,37 @@ func (c *consensusRuntime) calculateUptime(currentBlock *types.Header, epoch *ep
 		return nil
 	}
 
-	firstBlockInEpoch := calculateFirstBlockOfPeriod(currentBlock.Number, c.config.PolyBFTConfig.EpochSize)
-	lastBlockInPreviousEpoch := firstBlockInEpoch - 1
-
-	startBlock := (epoch.Number * c.config.PolyBFTConfig.EpochSize) - c.config.PolyBFTConfig.EpochSize + 1
-	endBlock := getEndEpochBlockNumber(epoch.Number, c.config.PolyBFTConfig.EpochSize)
-
-	blockHeader := currentBlock
-	validators := epoch.Validators
-
-	var found bool
+	blockExtra, err := GetIbftExtra(currentBlock.ExtraData)
+	if err != nil {
+		return nil, err
+	}
 
-	for blockHeader.Number > firstBlockInEpoch {
-		if err := calculateUptimeForBlock(blockHeader, validators); err != nil {
+	// calculate uptime for current epoch
+	for blockHeader.Number > epoch.FirstBlockInEpoch {
+		if err := calculateUptimeForBlock(blockExtra, epoch.Validators); err != nil {
 			return nil, err
 		}
 
-		blockHeader, found = c.config.blockchain.GetHeaderByNumber(blockHeader.Number - 1)
-		if !found {
-			return nil, blockchain.ErrNoBlock
-		}
+		blockHeader, blockExtra, err = c.getBlockData(blockHeader.Number - 1)
 	}
 
-	// since we need to calculate uptime for the last block of the previous epoch,
-	// we need to get the validators for the that epoch from the smart contract
-	// this is something that should probably be optimized
-	if lastBlockInPreviousEpoch > 0 { // do not calculate anything for genesis block
+	// calculate uptime for blocks from previous epoch that were not processed in previous uptime
+	// since we can not calculate uptime for the last block in epoch (because of parent signatures)
+	if blockHeader.Number > uptimeLookbackSize {
 		for i := 0; i < uptimeLookbackSize; i++ {
 			validators, err := c.config.polybftBackend.GetValidators(blockHeader.Number-2, nil)
 			if err != nil {
 				return nil, err
 			}
 
-			if err := calculateUptimeForBlock(blockHeader, validators); err != nil {
+			if err := calculateUptimeForBlock(blockExtra, validators); err != nil {
 				return nil, err
 			}
 
-			blockHeader, found = c.config.blockchain.GetHeaderByNumber(blockHeader.Number - 1)
-			if !found {
-				return nil, blockchain.ErrNoBlock
-			}
+			blockHeader, blockExtra, err = c.getBlockData(blockHeader.Number - 1)
 		}
 	}
 
-	epochID := epoch.Number
-
 	uptime := Uptime{EpochID: epochID}
```

Comment:
We can move this declaration down just before it's used

Reply:
Fix: https://github.com/0xPolygon/polygon-edge/pull/897/commits/7e3625bb48d731068c5c64aebc56aafe8d42ffa1